Update zabbix_host.py

tikoflano opened the pull request:
##### SUMMARY
Wrong usage of the "inventory_mode" makes the module always fail if the host has an inventory already. 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
zabbix_host.py

##### ADDITIONAL INFORMATION
The "inventory_mode" is part of the host object and not the inventory. (ref: https://www.zabbix.com/documentation/current/manual/api/reference/host/object)

tikoflano committed:
Update zabbix_host.py

The "inventory_mode" is part of the host object. (ref: https://www.zabbix.com/documentation/current/manual/api/reference/host/object)

D3DeFi commented:
Hello, not sure if this is the line you wanted to adjust.

When you request host details via Zabbix API as is the case for zabbix_host module:
```
zbx.host.get({'output': 'extend', 'selectInventory': 'extend', 'filter': {'host': [host_name]}})
```
You will get this in the response:
```
...
   "inventory" : {
      ...
      "inventory_mode" : "0",
      "hostid" : "20805",
...
```
Thus searching for inventory mode in `host['inventory']['inventory_mode']` should be valid.

I've tried to reproduce the behavior you're describing with Zabbix server 4.0, but without any luck. can you please provide more information such as traceback/error you are getting, Zabbix server version, example playbook that is failing for you?

tikoflano commented:
Now I realized that this is currently being addressed in the `devel` branch with line https://github.com/ansible/ansible/blob/4752547d35db67db38ac03bd0558c89d09b9487f/lib/ansible/modules/monitoring/zabbix/zabbix_host.py#L559

But I have AWX which uses currently ansible 2.8.5 on awx_task. On 2.9 isn't fixed yet so I'll have to wait for a new release. In the meantime I'll patch my own version.

--------------

It seems to be something that they changed specifically on 4.0. When I run the query to the API I get the response:

```
...
"inventory_mode": "1",
"inventory": {
    "type": "",
    "type_full": "",
    ...
...
```

Take a look at these docs and search for "inventory_mode":

[3.0](https://www.zabbix.com/documentation/3.0/manual/api/reference/host/object), [4.0](https://www.zabbix.com/documentation/4.0/manual/api/reference/host/object) and [4.4](https://www.zabbix.com/documentation/4.4/manual/api/reference/host/object)

On 4.0 they decided to create two separate variables for "inventory_mode", you **read** it form inside the `inventory` object and you write it inside the `host` object (IMO that's pretty weird). But they reverted it on 4.4 (4.2 works just like 4.0) to the previous way and also will be using that way on upcoming 5.0.

So, the module should be working just fine with version between (including) 4.0 and 4.2, but failing with 3.x and 4.4+. I'll update the PR with the correct version check in order to have it working with <4.0.

tikoflano committed:
Update zabbix_host.py

tikoflano committed:
Update zabbix_host.py

D3DeFi reviewed (approved):
Good catch, thank you! :) I think it is good to go. What do you think @sky-joker?

shipit

sky-joker commented on lib/ansible/modules/monitoring/zabbix/zabbix_host.py:
```diff
@@ -556,15 +556,16 @@ def check_all_properties(self, host_id, group_ids, status, interfaces, template_
-                if host['inventory']:
+            if host['inventory']:
+                # The way to read the inventory_mode is different on versions between [4.0.0, 4.4.0)
+                if LooseVersion(self._zbx_api_version) >= LooseVersion('4.0.0') and LooseVersion(self._zbx_api_version) < LooseVersion('4.4.0'):
```
```suggestion
                if LooseVersion(self._zbx_api_version) < LooseVersion('4.4.1'):
```

The following structure has been changed since version 4.4.1.

```
(snip)
    "inventory_mode": "0",
    "inventory": {
      "type": "",
      "type_full": "",
      "name": "",
      "alias": "fuga",
(snip)
```

So, please change the version condition here to less than 4.4.1.

Tested with the following version.

* 3.0.29
* 3.2.11
* 3.4.15
* 4.0.0
* 4.0.17
* 4.2.8
* 4.4.0
* 4.4.1
* 4.4.2
* 4.4.3
* 4.4.4
* 4.4.5
* 5.0.0alpha

tikoflano replied:
I checked the changelogs for [4.4.1](https://www.zabbix.com/rn/rn4.4.1) and indeed that was the version where it changed. But what about the condition `>= LooseVersion('4.0.0')`? According to the [docs ](https://www.zabbix.com/documentation/3.0/manual/api/reference/host/object) it worked like 4.4.1+

sky-joker replied:
Thank you for check Zabbix API documentation and my suggestion.  
I commented about `if` condition, please check it.

felixfontein commented:
Also, you need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

tikoflano committed:
Update zabbix_host.py

tikoflano commented:
> Also, you need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

Done.

tikoflano committed:
Create 66860-zabbix-api-inventory_mode-usage-fix.yml

tikoflano committed:
Update 66860-zabbix-api-inventory_mode-usage-fix.yml

sky-joker commented on lib/ansible/modules/monitoring/zabbix/zabbix_host.py:
```diff
@@ -556,15 +556,16 @@ def check_all_properties(self, host_id, group_ids, status, interfaces, template_
-                if host['inventory']:
+            if host['inventory']:
+                # The way to read the inventory_mode is different on versions between [4.0.0, 4.4.1)
+                if LooseVersion(self._zbx_api_version) >= LooseVersion('4.0.0') and LooseVersion(self._zbx_api_version) < LooseVersion('4.4.1'):
```
The following code should processing in Zabbix version 4.4.0 or less.

```
                    if int(host['inventory']['inventory_mode']) != self.inventory_mode_numeric(inventory_mode):
                        return True
```

But Zabbix version 3.0/3.2/3.4 isn't included following condition, so error occurs that when parameter updating.  
(Zabbix 3.0 is still in [support period](https://www.zabbix.com/life_cycle_and_release_policy) and needs to be supported by this module)

```
                if LooseVersion(self._zbx_api_version) >= LooseVersion('4.0.0') and LooseVersion(self._zbx_api_version) < LooseVersion('4.4.1'):
```

Could you please reconsider the above condition?

tikoflano replied:
Sory but I still don´t get it. There are two conditions:

- `if LooseVersion(self._zbx_api_version) >= LooseVersion('4.0.0') and LooseVersion(self._zbx_api_version) < LooseVersion('4.4.1'):`
    In this case check for `host['inventory']['inventory_mode']`. This covers Zabbix [4.0.0, 4.4.1)
- `else`:
    In this case check for `host['inventory_mode']`. This covers Zabbix (0.0.0, 4.0.0) and [4.4.1, 4.4.1+)

So as far I can tell your concern is about supporting version 3.0/3.2/3.4 which are covered by the `else` section because on those versions we need to check for `host['inventory_mode']` and not `host['inventory']['inventory_mode']`.

Is there something I'm missing (or maybe on the translation, English is not my first language, it is Spanish).

sky-joker replied:
Could you deploy Zabbix 3.0 in your test environment and run this [test playbook](https://gist.github.com/sky-joker/1c6f20f108000fa4bd236969d1c8e907) on Zabbix 3.0?  
Then, the following error occurs.(**KeyError: 'inventory_mode'**)

```
(venv) [root@localhost ~]# ansible-playbook pr_test_66860_for_zabbix.yml -vvv
(snip)
TASK [update inventory_zabbix param for host] **********************************************************************************************************************************
task path: /root/dev/PR/66860/pr_test_66860_for_zabbix.yml:53
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482 `" && echo ansible-tmp-1582966175.1346583-161015819237482="` echo /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482 `" ) && sleep 0'
Using module file /root/dev/PR/66860/ansible-1/lib/ansible/modules/monitoring/zabbix/zabbix_host.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-18759uhn0vly3/tmp83nxyjpl TO /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/ /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/root/dev/PR/66860/ansible-1/venv/bin/python /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py", line 102, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.monitoring.zabbix.zabbix_host', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py", line 925, in <module>
  File "/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py", line 885, in main
  File "/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py", line 565, in check_all_properties
KeyError: 'inventory_mode'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible.modules.monitoring.zabbix.zabbix_host', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py\", line 925, in <module>\n  File \"/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py\", line 885, in main\n  File \"/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py\", line 565, in check_all_properties\nKeyError: 'inventory_mode'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

**Zabbix 3.0 install procedure**

https://www.zabbix.com/jp/download?zabbix=3.0&os_distribution=centos&os_version=7&db=mysql&ws=apache


Thank you for in advance.

tikoflano replied:
I don't have an environment to test it, but I see that you already test it against all supported versions so I changed the test.

sky-joker replied:
Thank you @tikoflano  
I checked each Zabbix version source code for the host data JSON structure.  
Source code the following.

* [3.0](https://github.com/zabbix/zabbix/blob/aa382ebff0fe2a7da2d92f72bbc861327d6b6e3c/frontends/php/host_prototypes.php#L276
)
* [4.0](https://github.com/zabbix/zabbix/blob/98553278ba431b8d574187d96d4fad6c4dd4717b/frontends/php/host_prototypes.php#L287
)
* [4.2](https://github.com/zabbix/zabbix/blob/acdcb4e70e30fdd96d6d1c741c2b061c7b60c222/frontends/php/host_prototypes.php#L287
)
* [4.4.0](https://github.com/zabbix/zabbix/blob/cfac660b256d766a8bfc6ba69e5d76c8b4e627bd/frontends/php/host_prototypes.php#L287)
* [4.4.1](https://github.com/zabbix/zabbix/blob/8870606e6a83cd81c979ac1f168645a60ffc6149/frontends/php/host_prototypes.php#L284
)

Looking at the source code, you can see that the JSON structure for return value changed from 4.4.1.  
I want to merge this PR, but devel branch is freeze, so merging is blocked now.  
see: https://github.com/ansible-collections/overview/blob/master/README.rst#timeline

I think probably the Zabbix module moves to `community.general`, but I don't know when `community.general` is created.  
If this PR to merge that I think you need to reopen this PR there, after when `community.general` created.

tikoflano committed:
Update zabbix_host.py

Even though the docs says different, sky-joker tested this condition against all supported versions and it passed.

tikoflano committed:
Update zabbix_host.py

Even though the docs says different, @sky-joker tested this condition against all supported versions and it passed.

tikoflano committed:
Merge branch 'patch-1' of https://github.com/tikoflano/ansible-1 into patch-1

tobiasehlert reviewed (approved):
This change worked fine for me to run :)

D3DeFi commented:
Unfortunately, your contribution can no longer be merged to [ansible/ansible](https://github.com/ansible/ansible). The way how community content is being shipped within Ansible is [changing](https://docs.ansible.com/ansible/devel/user_guide/collections_using.html#collections) as of right now with the preparations for Ansible 2.10. 

All development regarding Zabbix content has been moved to a separate [ansible-collections/community.zabbix](https://github.com/ansible-collections/community.zabbix) repository. Please resubmit your PR against the new repository with the changes we have requested in the previous comments and reviews (if any).

This separation now allows us to address your contribution much quicker than before. We are giving each contributor a few weeks (roughly a month) notice to move his/hers PRs to a new repository.

After a month, we will try to migrate all leftover PRs by ourselves, but author information may get lost in this way.

Sorry for the inconvenience and thank you for the understanding.

close_me
